src/main.rs

```diff
@@ -63,44 +64,65 @@ fn update_stats(start: Instant, typed: usize, attempts: usize) {
     );
 }
 
-struct Attempter {
-    stdout: RawTerminal<io::Stdout>,
+struct WindowHandler {
     practice_set: PracticeSet,
-    chunk: VecDeque<String>,
+    chunk_vec: VecDeque<VecDeque<char>>,
+    window_len: usize,
 }
 
-impl Attempter {
-    fn new(mut practice_set: PracticeSet) -> Self {
-        Attempter {
-            chunk: practice_set
-                .choose_n(10)
+impl WindowHandler {
+    fn new(mut practice_set: PracticeSet, window_len: usize) -> Self {
+        WindowHandler {
+            chunk_vec: practice_set
+                .choose_n((window_len / 2) + 1)
                 .into_iter()
-                .map(|word| word.chars().rev().collect())
+                .map(|word| word.chars().collect())
                 .collect(),
             practice_set,
-            stdout: io::stdout().into_raw_mode().unwrap(),
+            window_len,
         }
     }
 
-    fn run(&mut self) -> Result<usize, ()> {
-        let chunk: String = self
-            .chunk
+    fn chunk_view(&self) -> String {
+        self.chunk_vec
             .iter()
-            .flat_map(|w| w.chars().rev().chain(std::iter::once(' ')))
-            .take(20)
-            .collect();
+            .flat_map(|word| word.iter().copied().chain(std::iter::once(' ')))
+            .take(self.window_len)
+            .collect()
+    }
 
-        let goal = match self.chunk[0].pop() {
+    // Return next character to type
+    fn next_char(&mut self) -> char {
+        match self.chunk_vec[0].pop_front() {
             Some(c) => c,
             None => {
-                self.chunk.pop_front();
-                self.chunk
-                    .push_back(self.practice_set.choose().chars().rev().collect());
+                self.chunk_vec.pop_front();
+                let next_word: VecDeque<_> = self.practice_set.choose().chars().collect();
```

Comment:
is the type annotation really necessary?

Reply:
🙄